src/buildercore/cfngen.py

```diff
@@ -109,108 +120,169 @@ def build_context(pname, **more_context): # pylint: disable=too-many-locals
         #}
     }
 
-    context = copy.deepcopy(defaults)
+    context = deepcopy(defaults)
     context.update(more_context)
 
+    ensure('stackname' in context, "'stackname' not provided") # this still sucks
+
     # proceed with wrangling
 
-    # TODO: don't like this. if a stackname is required, make it a parameter.
-    # stackname used to be derived inside this func from pname + id + cluster number
-    ensure(context['stackname'], "a stackname wasn't provided.")
-    stackname = context['stackname']
+    # why shuffle? so we don't have a context builder depending
+    # on another context builder mutating the context in a specific way
+    # wrangler_list = shuffle([
+    wrangler_list = [
+        partial(build_context_rds, existing_context=existing_context),
+        build_context_aws,
+        build_context_ec2,
+        build_context_elb,
+        build_context_cloudfront,
+        build_context_sns_sqs,
+        build_context_s3,
+        build_context_cloudfront,
+        build_context_fastly,
+        build_context_gcp,
+        build_context_subdomains,
+        build_context_elasticache,
+        build_context_vault,
+    ]
+
+    # exceptions to the rule ...
+    wrangler_list = [project_wrangler] + wrangler_list
+
+    for wrangler in wrangler_list:
+        # so functions can't modify the context in-place or
+        # reference a bit of project_data and then change it
+        context = wrangler(deepcopy(project_data), deepcopy(context))
+
+    return context
+
+def build_context_aws(pdata, context):
+    if 'aws' not in pdata:
+        return context
+    keepers = [
+        'region',
+        'vpc-id',
+        'subnet-id',
```

Comment:
maybe the `subnet-*` and `redundant-subnet-*` fields could be moved to `context['ec2']` (not part of this PR)
RDS refers to them only as "allows EC2 instances to connect to me"; not sure if ElastiCache is the same.

Reply:
`context['elasticache']` has its own `subnets` field so yes, we'll move them.
`region` and `vpc-id` are so general for AWS that they are already in the right place.

Reply:
`context['elb']` will need to get them from `aws.elb.subnets`, as it doesn't at the moment

Reply:
:+1: